### PR TITLE
Implement class DynamicCreative 

### PR DESCRIPTION
**Pull request by MKFMIKU**
I change the origin canvas in order to display initial elements to correct positions

### CODE REVIEW

File: lib/DynamicCreative.js [hunk 1]
```diff
@@ -1,42 +1,99 @@
...
+        this.elements = elements
+    }
 
-}
+    /**
+     * Resize the canvas
+     * @param {Object} canvas
+     * @return {Object}
+     */
+    toParams(canvas) {
```
Comment (yinrong) [inline lib/DynamicCreative.js]: any function should not exceed one page.

File: lib/DynamicCreative.js [hunk 2]
```diff
@@ -1,42 +1,99 @@
...
+             * Real question
+             */
+            ans.map(item => {
+                if (item.type == 'Image') {
+                    var tempW
+                    if ((item.w / this.origin.w) > 0.4) {
+                        /**
+                         * Assume the image as a full
+                         */
+                        var tempW = canvas.w * 0.8
```
Comment (yinrong) [inline lib/DynamicCreative.js]: variable naming convention: foo_bar

File: lib/DynamicCreative.js (hunk 1 above)
Comment (yinrong) [reply lib/DynamicCreative.js]: use a standalone class is better

File: lib/DynamicCreative.js [hunk 3]
```diff
@@ -1,42 +1,99 @@
...
 
-}
+    /**
+     * Resize the canvas
+     * @param {Object} canvas
+     * @return {Object}
+     */
+    toParams(canvas) {
+        var ans = [].concat(JSON.parse(JSON.stringify(this.elements)))
+        ans = ans.sort(function(a, b) {
```
Comment (yinrong) [inline lib/DynamicCreative.js]: use arrow function

File: lib/DynamicCreative.js (hunk 1 above)
Comment (yinrong) [reply lib/DynamicCreative.js]: screen shot 2017-05-17 at 10 58 09

Comment (yinrong) [reply lib/DynamicCreative.js]: it seems you need a better algorithm.

File: lib/DynamicCreative.js [hunk 4]
```diff
@@ -1,42 +1,99 @@
...
 
-module.exports = DynamicCreative
 
+        if (scale > 1) {
+            var move_x = (canvas.w - this.origin.w) / 2
+            ans.map(item => {
+                item.x += move_x
+            })
+
+        } else if (scale > 0.6) {
```
Comment (yinrong) [inline lib/DynamicCreative.js]: why different scale have different logic? the code is not easy to understand

File: lib/DynamicCreative.js [hunk 5]
```diff
@@ -1,42 +1,99 @@
...
+            ans.map(item => {
+                if (item.type == 'Image') {
```
Comment (yinrong) [inline lib/DynamicCreative.js]: use '===' and '!==' in most cases

File: lib/DynamicCreative.js [hunk 6]
```diff
@@ -1,42 +1,99 @@
...
+    }
 
-}
+    /**
+     * Resize the canvas
+     * @param {Object} canvas
+     * @return {Object}
+     */
+    toParams(canvas) {
+        var ans = [].concat(JSON.parse(JSON.stringify(this.elements)))
```
Comment (yinrong) [inline lib/DynamicCreative.js]: do NOT use abbreviation unless it absolutely have no ambiguity.

File: lib/DynamicCreative.js [hunk 7]
```diff
@@ -1,42 +1,99 @@
...
+                                tempW = ans[i].x - item.x - 10
+                                item.h = item.h * (tempW / item.w)
+                                item.w = tempW
+                            }
+                        }
+                    } else {
+                        item.w = canvas.w * 0.8
+                        item.x = canvas.w * 0.1
+                    }
+                }
```
Comment (yinrong) [inline lib/DynamicCreative.js]: please try to avoid too many level of indents.

File: lib/DynamicCreative.js (hunk 1 above)
Comment (MKFMIKU) [reply lib/DynamicCreative.js]: Whether I should keep the origin order？I think it's a important information. @yinrong

Comment (yinrong) [reply lib/DynamicCreative.js]: that's your definition and your decision. in the origin problem, we say an empty area is bad.

File: lib/DynamicCreative.js [hunk 8]
```diff
@@ -1,42 +1,99 @@
...
-        h: 128 },
-      { id: 'text2',
-        type: 'Text',
-        x: 24,
-        y: 284.6666666666667,
-        w: 192,
-        h: 45 },
-    ]
-  }
+    constructor(canvas) {
```
Comment (yinrong) [inline lib/DynamicCreative.js]: please also use 2 spaces per level of indent

**Comment by MKFMIKU** [comment]
@yinrong Sorry, I cannot find a good algorithm to fit the square into proper place

**Comment by MKFMIKU** [comment]
I use new algorithm and change the background color of html
qq 20170518085048

File: lib/DynamicCreative.js [hunk 9]
```diff
@@ -1,42 +1,99 @@
...
 
-}
-
-module.exports = DynamicCreative
+  /**
+   * If scale<1, roate the canvas into origin
+   * @param {Object} canvas
+   * @param {Array} ans
+   */
+  rotate_ans(canvas, ans) {
```
Comment (yinrong) [inline lib/DynamicCreative.js]: naming convention:
```
publicMethod
_privateMethod
_privateClassMember
```

**Review by yinrong — COMMENTED** [review]
There are two main problems:

1. Text should remain horizontal. 
2. There are complete NO order now, and this not good.

**Comment by MKFMIKU** [comment]
🐛 Give me some time, I will try to keep the Text horizontal.
And tell the truth, I have kept the order but the algorithm split the Image and the Text.

**Comment by MKFMIKU** [comment]
I cannot figure out how to solve it in a perfect way 😞 
Would you like to give me some tips? 😆 

@yinrong

**Comment by yinrong** [comment]
use different rules if you can't get it done in one

**Comment by MKFMIKU** [comment]
The horizontal text not really horizontal, they just too thin, I change some parameter to let them be fat. 
Use different rule will cause other unexpected bug. 😞 

image

**Comment by yinrong** [comment]
image

**Comment by yinrong** [comment]
The above picture shows what is a good result.

**Comment by MKFMIKU** [comment]
The task is so FUN and complex, I can't believe that it appears in a full stack interview. 🤣 🤣 🤣 
image

**Pull request closed without merging**